### ISSUE TITLE
 climada.util.coordinates.lon_normalize crashes tc_tracks.calc_perturbed_trajectories

### ISSUE THREAD

**Issue by chahank** [CLIMADA-project/climada_python#148]
The changes to ` climada.util.coordinates.lon_normalize` and `tc_tracks.calc_perturbed_trajectories` are unstable. The following code now fails:

```
from climada.hazard import TCTracks
basin ='NA'
tc_tracks = TCTracks()
tc_tracks.read_ibtracs_netcdf(basin=basin, year_range=(1950, 2020)) 
tc_tracks.equal_timestep(time_step_h=1)
#Need to remove tracks with a single time step for calc_perturbed_trajectories
filter_only_one_time = [track for track in tc_tracks.data if track.time.size >= 2]
tc_tracks2= TCTracks()
tc_tracks2.data = filter_only_one_time
tc_tracks2.calc_perturbed_trajectories(nb_synth_tracks=10)
```

Error msg from `tc_tracks2.calc_perturbed_trajectories(nb_synth_tracks=10)`: climada.util.coordinates - WARNING - lon_normalize: killed before finishing

The `tc_tracks2` object is then empty.

Note that his bug was not present on the 25.01.2020.

**Comment by bguillod** [comment]
First of all, does the following work?
```
tc_tracks.calc_perturbed_trajectories(nb_synth_tracks=10)
```
(it probably shouldn't).

My guess is that the call to `lon_normalize` comes from the following lines in `_one_rnd_walk`:
```
        angular_dist = climada.util.coordinates.dist_approx(i_track.lat.values[:-1, None],
                                                            i_track.lon.values[:-1, None],
                                                            i_track.lat.values[1:, None],
                                                            i_track.lon.values[1:, None],
                                                            method="geosphere",
                                                            units="degree")[:, 0, 0]
```
because `lon_normalize` is called in `dist_approx`.

Could you identify what changed in dist_approx and in lon_normalize since 25.01.2020 (I hope you actually mean 25.01.2021)? That should give you part of the answer, hopefully.

**Comment by chahank** [comment]
`tc_tracks.calc_perturbed_trajectories(nb_synth_tracks=10)` does not work because some of the tracks have only one time step and then the method complains.

**Comment by tovogt** [comment]
Since calc_perturbed_trajectories iterates over all tc tracks, this bug should be independent of the number of tracks in your `TCTracks` object. Would you please try to restrict your example to a single track that causes the error message?

**Comment by chahank** [comment]
Ok, I will try to see.

**Comment by tovogt** [comment]
Sorry, I was wrong: The procedure does depend on the number of tracks due to the randomness involved in the process. Still, I was able to find a smaller setup that reproduces your problem:
```
basin ='NA'
tc_tracks = TCTracks()
tc_tracks.read_ibtracs_netcdf(basin=basin, year_range=(1950, 1986))
tc_tracks.data = [track for track in tc_tracks.data if track.time.size >= 2]
tc_tracks.equal_timestep(time_step_h=1)
tc_tracks.calc_perturbed_trajectories(nb_synth_tracks=10)
```
I found that the problem isn't directly related to `lon_normalize`. The problem is that one hurricane in the ensemble reaches high latitudes (http://ibtracs.unca.edu/index.php?name=v04r00-1986226N30276) and a very unfortunate choice of random parameters lets this track continue on until it almost reaches the north pole (latitude 89° north) where it starts to spin around the pole like crazy so that longitude is increased to more than 4000° in `_one_rnd_walk`. Finally, the `lon_normalize` function refuses to operate on that kind of longitudinal value because it looks corrupt.

I'm not really familiar with the algorithm behind `calc_perturbed_trajectories` but this example looks like a bug in the algorithmic design because tropical cyclones should never reach that far north. However, if you only need a quick fix, just make sure to renormalize longitudinal values to [-180, 180] after taking a step in `_get_destination_points`:
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L430-L431

**Comment by bguillod** [comment]
Thanks @tovogt for looking into this, very helpful. Some comments below.

> I found that the problem isn't directly related to `lon_normalize`. The problem is that one hurricane in the ensemble reaches high latitudes (http://ibtracs.unca.edu/index.php?name=v04r00-1986226N30276) and a very unfortunate choice of random parameters lets this track continue on until it almost reaches the north pole (latitude 89° north) where it starts to spin around the pole like crazy so that longitude is increased to more than 4000° in `_one_rnd_walk`. Finally, the `lon_normalize` function refuses to operate on that kind of longitudinal value (larger than 3600°) because it looks corrupt.

That is indeed unfortunate and should not happen.

> I'm not really familiar with the algorithm behind `calc_perturbed_trajectories` but this example looks like a bug in the algorithmic design because tropical cyclones should never reach that far north. However, if you only need a quick fix, just make sure to renormalize longitudinal values to [-180, 180] after taking a step in `_get_destination_points`

Agreed, this would require a change in the algorithm, setting up some bounds to latitude or decreasing wind speed/pressure significantly if those bounds are being excessed.
A side-remark: could someone check if when looking at tracks crossing the anti-meridian the longitude values are correctly within -180/+180 degrees? I guess they are, but if not we indeed need to add a call to `lon_normalize` there.

> Since I'm not familiar with `calc_perturbed_trajectories`, I will not continue to work on this issue. I only looked into it because I'm the author of `lon_normalize`.

Sure. I am familiar with this function, but I do not have a concrete way how to do this as it is related to TC dynamics (also, I don't have time now). It should be done by someone more familiar with this topic than myself.

**Comment by chahank** [comment]
@tovogt  Thank you for the detective work!

@bguillod Could you address this in your update #152 ? Adding a hard limit for synthetic tracks at for example 61 or 70latitude would do the trick.

**Comment by bguillod** [comment]
> @bguillod Could you address this in your update #152 ? Adding a hard limit for synthetic tracks at for example 61 or 70latitude would do the trick.

Yes and no. We would need to define the specifics of what happens when the hard limit is reached. If I had time (which I don't at the moment) I'd probably do the following:
1. Look at how wind speed, pressure, and all other variables evolve (decay) when they move towards high latitude, and try to estimate if I can fit an (exponential?) decay, for instance by latitudinal bands (or with latitude as an additional parameter).
2. Apply this decay as an additional step in the synthetic tracks generation (a bit like the landfall decay that is currently implemented).
3. If the track parameters (esp. pressure) reach a point below a certain value (e.g., become lower than TC category 1), I would end the track there.

However this (especially step 1) takes some time. Would you @chahank by any chance have time to do this analysis, or know someone who could?

Of course, whether it is worth investing this time is tightly linked to whether the current synthetic tracks algorithm is kept or is replaced by something else.

**Comment by bguillod** [comment]
The quick fix simply cuts off tracks when they move outside of the latitudinal range [-70, 70]. If TC category is > 1 at the cutoff point, it issues a warning.
